File: 0016-3sum-closest/0016-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        if(nums == null || nums.length < 3){
            return 0;
        }

        Arrays.sort(nums);
        int closest = nums[0] + nums[1]+ nums[2];

        for(int i = 0; i < nums.length; i++){
            int j = i+1;
            int k = nums.length-1;
            while(j < k){
                int sum = nums[i] + nums[j] + nums[k];
                if(sum == target){
                    return target;
                }
                if (Math.abs(target - sum) < Math.abs(target - closest)) {
                    closest = sum;
                }

                if(sum < target){
                    j++;
                } else {
                    k--;
                }
            }
        }

        return closest; 
    }
}

File: 0016-3sum-closest/0016-3sum-closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if nums is None or len(nums) < 3:
            return -1

        nums.sort()
        closest = nums[0] + nums[1] + nums[2]

        for i in range(len(nums)):
            j = i+1
            k = len(nums)-1
            while j < k:
                sum =  nums[i] + nums[j] + nums[k]
                if(sum == target):
                    return target
                if(abs(target-sum) < abs(target-closest)):
                    closest = sum
                
                if sum < target:
                    j = j+1
                else:
                    k = k-1
    
        return closest
        